### PR TITLE
Restore ability to share sharedlibs in create_test.

### PR DESCRIPTION
**Pull request by jgfouca**
Change list:
1) Split BUILD phase into MODEL_BUILD and SHAREDLIB_BUILD
2) case.build now builds all models in parallel!
3) Sharedlib builds are serialized by clever use of procs_needed
4) Crucial Makefile fix from Jim Edwards
5) Add --model-only, --sharedlib-only options to case.build
6) Add --model-only, --sharedlib-only options to case.test_build
7) Add support for model-only, sharedlib-only to all the SystemTest classes
8) New tools for debugging: case_diff, normalize_cases.
 - Used to find subte differences in infrastructure behavior at the case level

9) Move most of check_input_files into a library
10) Remove non-threadsafe things from preview_namelists
11) Encapsulate fake system tests into a common base class
12) Add some missing fake system tests (TESTBUILDFAIL, TESTRUNSLOWPASS)
13) Add some timing data to case.build to assist future optimizations
14) Add prototype replacement for preview_namelists (not yet active)

### CODE REVIEW

**Comment by jgfouca** [comment]
Going to add annotations for you.

File: scripts/Tools/check_input_data [hunk 1]
```diff
@@ -62,91 +57,6 @@ formatter_class=argparse.ArgumentDefaultsHelpFormatter
     return args.svn_loc, args.input_data_root, args.data_list_dir, args.download
 
 ###############################################################################
```
Comment (jgfouca) [inline scripts/Tools/check_input_data]: Moved most of this to a library so other tools can call check_input_data without having to fork a subprocess.

File: utils/python/CIME/SystemTests/nck.py [hunk 2]
```diff
@@ -41,21 +40,21 @@ def build(self):
                         self._case.set_value("ROOTPE_%s"%comp, "%s"%int(rootpe/2))
             self._case.flush()
 
-
             run_cmd("case.setup -clean -testmode")
```
Comment (jgfouca) [inline utils/python/CIME/SystemTests/nck.py]: I had to undo some of your changes to this file in order to get identical behavior in both one-phase and two-phase builds. Might be worth thinking about those changes and if we need to bring them back in some form.

File: utils/python/CIME/build.py [hunk 3]
```diff
@@ -498,3 +525,14 @@ def build_libraries(exeroot, caseroot, cimeroot, libroot, sharedlibroot, compile
...
 
 ###############################################################################
+def _build_model_thread(config_dir, caseroot, bldroot, compspec, file_build,
+                        exeroot, model, comp, objdir, incroot, thread_bad_results):
+###############################################################################
+    stat = run_cmd("%s/buildlib %s %s %s >> %s 2>&1" %
+                   (config_dir, caseroot, bldroot, compspec, file_build),
+                   from_dir=os.path.join(exeroot, model), ok_to_fail=True)[0]
+    if (stat != 0):
+        thread_bad_results.append("ERROR: %s.buildlib failed, see %s" % (comp, file_build))
```
Comment (jgfouca) [inline utils/python/CIME/build.py]: Append is atomic.

File: utils/python/CIME/system_test.py [hunk 4]
```diff
@@ -149,12 +150,12 @@ def __init__(self, test_names,
                                                                self._baseline_gen_name)
 
             # Compute baseline_root
-            if baseline_root is None:
-                self._baseline_root = self._machobj.get_value("CCSM_BASELINE")
-            else:
-                self._baseline_root = baseline_root
+            self._baseline_root = baseline_root if baseline_root is not None \
```
Comment (jgfouca) [inline utils/python/CIME/system_test.py]: I think this is a reasonable way to split a ternary across two lines if we must.

File: utils/python/CIME/utils.py [hunk 5]
```diff
@@ -149,8 +149,7 @@ def run_cmd(cmd, ok_to_fail=False, input_str=None, from_dir=None, verbose=None,
     if (arg_stderr is _hack):
         arg_stderr = subprocess.PIPE
```
Comment (jgfouca) [inline utils/python/CIME/utils.py]: May need to discuss this further. If the verbose argument is True but logging.info is off, then logging.info prints nothing when it should. If the opposite is true, then we have the opposite problem.

**Comment by jgfouca** [comment]
Annotations complete.

File: utils/python/CIME/build.py [hunk 6]
```diff
@@ -274,25 +285,41 @@ def case_build(caseroot, testmode):
...
     env_module = EnvModule(mach, compiler, cimeroot, caseroot, mpilib, debug)
     env_module.load_env_for_case()
-    # Need to flush case xml to disk before calling preview_namelists
-    case.flush()
-    run_cmd("./preview_namelists")
 
-    build_checks(case, build_threaded, comp_interface, use_esmf_lib, mpilib,
+    if not sharedlib_only:
+        # Need to flush case xml to disk before calling preview_namelists
+        case.flush()
```
Comment (jedwards4b) [inline utils/python/CIME/build.py]: I think that this case.flush() needs to happen in both passes - the case may have been updated with something needed for the sharedlib build.

File: utils/python/CIME/check_input_data.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        else:
+            return True
+
+def check_input_data(case=None, svn_loc=None, input_data_root=None, data_list_dir="Buildconf", download=False):
+    """
+    Return True if no files missing
+    """
+    # Fill in defaults as needed
+    case = Case() if case is None else case
+    svn_loc = SVN_LOCS[get_model()] if svn_loc is None else snv_loc
```
Comment (jedwards4b) [inline utils/python/CIME/check_input_data.py]: typo here: snv_lock should be svn_lock

File: utils/python/CIME/check_input_data.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+            line = line.strip()
+            if (line and not line.startswith("#")):
+                tokens = line.split('=')
+                description, full_path = tokens[0].strip(), tokens[1].strip()
+                if(full_path):
+                # expand xml variables
+                    full_path = case.get_resolved_value(full_path)
+                    rel_path  = full_path.replace(input_data_root, "")
+
+                    if (not os.path.exists(full_path)):
```
Comment (jedwards4b) [inline utils/python/CIME/check_input_data.py]: There are some special values of rel_path that we need to ignore - some of the component models set things like
'NULL' or 'same_as_TS' - basically if rel_path does not contain '/' (a directory tree) you can assume it's a special value and ignore it (perhaps with a warning)

File: scripts/Tools/check_input_data [hunk 9]
```diff
@@ -62,91 +57,6 @@ formatter_class=argparse.ArgumentDefaultsHelpFormatter
...
-                else:
-                    model = os.path.basename(data_list_file).split('.')[0]
-                    logging.warning("Model %s no file specified for %s"%(model,description))
-
-    return no_files_missing
-
-###############################################################################
 def _main_func(description):
 ###############################################################################
     if ("--test" in sys.argv):
```
Comment (jedwards4b) [inline scripts/Tools/check_input_data]: For some reason github won't let me annotate line 68.   But the arguments to check_input_data are wrong here - should have a case first.

File: utils/python/CIME/utils.py (hunk 5 above)
Comment (jedwards4b) [inline utils/python/CIME/utils.py]: I plan to rework the logging facility a bit and will address this at that point.

**Comment by jedwards4b** [comment]
Review is complete - a few minor changes needed.

File: utils/python/CIME/build.py (hunk 6 above)
Comment (jgfouca) [inline utils/python/CIME/build.py]: Done. Can't hurt.

File: utils/python/CIME/check_input_data.py (hunk 7 above)
Comment (jgfouca) [inline utils/python/CIME/check_input_data.py]: Done. Must have forgotten to run pylint on this file.

File: utils/python/CIME/check_input_data.py (hunk 8 above)
Comment (jgfouca) [inline utils/python/CIME/check_input_data.py]: Done.

**Pull request merged**